File: languages/rust/experiments/java_rust/generics.java
class Generics {
    public static void main(String[] args) {
        var containerInt = new Container<Integer>(456);
        var containerString = new Container<String>("Yolo");

        System.out.println("Compare with other: " + containerInt.compareItem(11));
        System.out.println("And more: " + Generics.compare("aab", "bba"));
    }

    public static <T extends Comparable> int compare(T item1, T item2) {
        return item1.compareTo(item2);
    }
}

class Container<T extends Comparable> {
    T item;

    Container(T t) {
        this.item = t;
    }

    int compareItem(T other) {
        return item.compareTo(other);
    }
}

File: languages/rust/experiments/java_rust/generics.rs
use std::cmp::{Ord, Ordering};

fn main() {
    let container_int = Container { item: 456 };
    let container_string = Container {
        item: "Hello".to_string(),
    };

    println!("Compare int: {:?}", container_int.compare_item(&123));
    println!("Generic compare: {:?}", compare_items(&223, &332));
}

struct Container<T: Ord> {
    item: T,
}

impl<T: Ord> Container<T> {
    fn compare_item(&self, other: &T) -> Ordering {
        self.item.cmp(other)
    }
}

fn compare_items<T: Ord>(item1: &T, item2: &T) -> Ordering {
    item1.cmp(item2)
}
